File: apps/ios/Tests/CloudflareAccessTestTokens.swift
import Foundation
import Security
import Testing
@testable import OpenClaw

/// Runtime signing exercises the platform verifier without storing a private signing key in fixtures.
struct CloudflareAccessTestTokens {
    let key: SecKey
    let jwks: Data

    init() throws {
        self.key = try #require(SecKeyCreateRandomKey([
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeySizeInBits as String: 2048,
        ] as CFDictionary, nil))
        let publicKey = try #require(SecKeyCopyPublicKey(self.key))
        let der = try #require(SecKeyCopyExternalRepresentation(publicKey, nil)) as Data
        var outer = Array(der)[...]
        var sequence = try Self.readDER(&outer, tag: 0x30)[...]
        var modulus = try Self.readDER(&sequence, tag: 0x02)
        if modulus.first == 0 { modulus.removeFirst() }
        let exponent = try Self.readDER(&sequence, tag: 0x02)
        self.jwks = try JSONSerialization.data(withJSONObject: ["keys": [[
            "kty": "RSA", "kid": "test-key", "alg": "RS256", "use": "sig",
            "n": Self.encode(Data(modulus)), "e": Self.encode(Data(exponent)),
        ]]])
    }

    static func application() throws -> CloudflareAccessApplication {
        try CloudflareAccessApplication(
            origin: CloudflareAccessOrigin(#require(URL(string: "https://gateway.example.test:8443"))),
            issuer: #require(URL(string: "https://example.cloudflareaccess.com")),
            audience: "test-audience")
    }

    func token(_ claims: [String: Any], algorithm: String = "RS256") throws -> String {
        let header = try JSONSerialization.data(withJSONObject: ["alg": algorithm, "kid": "test-key"])
        let payload = try JSONSerialization.data(withJSONObject: claims)
        let message = "\(Self.encode(header)).\(Self.encode(payload))"
        let signature = try #require(SecKeyCreateSignature(
            self.key, .rsaSignatureMessagePKCS1v15SHA256, Data(message.utf8) as CFData, nil)) as Data
        return "\(message).\(Self.encode(signature))"
    }

    func session(subject: String = "test-subject", expires: Date = Date().addingTimeInterval(3600)) throws
        -> CloudflareAccessSession
    {
        let application = try Self.application()
        let token = try self.token([
            "iss": application.issuer.absoluteString, "aud": [application.audience],
            "type": "app", "sub": subject, "exp": expires.timeIntervalSince1970,
        ])
        return CloudflareAccessSession(application: application, subject: subject, token: token, expiresAt: expires)
    }

    private static func encode(_ data: Data) -> String {
        data.base64EncodedString().replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
    }

    private static func readDER(_ bytes: inout ArraySlice<UInt8>, tag: UInt8) throws -> [UInt8] {
        let actualTag = bytes.popFirst()
        #expect(actualTag == tag)
        let lengthByte = bytes.popFirst()
        let first = try #require(lengthByte)
        var length = Int(first)
        if first >= 128 {
            length = 0
            for _ in 0..<(first & 0x7F) {
                let next = bytes.popFirst()
                length = try length * 256 + Int(#require(next))
            }
        }
        #expect(bytes.count >= length)
        let value = Array(bytes.prefix(length))
        bytes = bytes.dropFirst(length)
        return value
    }
}

File: apps/android/app/src/test/java/ai/openclaw/app/gateway/CloudflareAccessTestTokens.kt
package ai.openclaw.app.gateway

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.math.BigInteger
import java.security.KeyPairGenerator
import java.security.Signature
import java.security.interfaces.RSAPublicKey
import java.util.Base64

internal object CloudflareAccessTestTokens {
  val application = CloudflareAccessApplication(CloudflareAccessOrigin.from("https://gateway.example.test:8443"), CloudflareAccessJWT.issuer("example.cloudflareaccess.com"), "test-audience")
  private val key = KeyPairGenerator.getInstance("RSA").apply { initialize(2048) }.generateKeyPair()
  private val publicKey = key.public as RSAPublicKey

  private fun unsigned(value: BigInteger): ByteArray = value.toByteArray().let { if (it.first() == 0.toByte()) it.copyOfRange(1, it.size) else it }

  private fun encode(value: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(value)

  val jwks =
    JsonObject(
      mapOf(
        "keys" to
          JsonArray(
            listOf(
              JsonObject(
                mapOf(
                  "kty" to JsonPrimitive("RSA"),
                  "kid" to JsonPrimitive("test-key"),
                  "alg" to JsonPrimitive("RS256"),
                  "use" to JsonPrimitive("sig"),
                  "n" to JsonPrimitive(encode(unsigned(publicKey.modulus))),
                  "e" to JsonPrimitive(encode(unsigned(publicKey.publicExponent))),
                ),
              ),
            ),
          ),
      ),
    ).toString().toByteArray()

  fun token(
    claims: JsonObject,
    algorithm: String = "RS256",
  ): String {
    val header = JsonObject(mapOf("alg" to JsonPrimitive(algorithm), "kid" to JsonPrimitive("test-key")))
    val message = "${encode(header.toString().toByteArray())}.${encode(claims.toString().toByteArray())}"
    val signature =
      Signature
        .getInstance("SHA256withRSA")
        .apply {
          initSign(key.private)
          update(message.toByteArray())
        }.sign()
    return "$message.${encode(signature)}"
  }

  fun claims(
    subject: String = "test-subject",
    expires: Double = System.currentTimeMillis() / 1000.0 + 3600,
  ): JsonObject =
    JsonObject(
      mapOf(
        "iss" to JsonPrimitive(application.issuer.toString()),
        "aud" to JsonArray(listOf(JsonPrimitive(application.audience))),
        "type" to JsonPrimitive("app"),
        "sub" to JsonPrimitive(subject),
        "exp" to JsonPrimitive(expires),
      ),
    )

  fun session(
    subject: String = "test-subject",
    expires: Double = System.currentTimeMillis() / 1000.0 + 3600,
  ): CloudflareAccessSession = CloudflareAccessSession(application, subject, expires, token(claims(subject, expires)))

  fun metadata(hostname: String = application.origin.uri.host): String =
    token(
      JsonObject(
        mapOf(
          "type" to JsonPrimitive("match"),
          "hostname" to JsonPrimitive(hostname),
          "auth_domain" to JsonPrimitive(application.issuer.host),
          "aud" to JsonPrimitive(application.audience),
          "iat" to JsonPrimitive(System.currentTimeMillis() / 1000.0),
        ),
      ),
    )
}
